content/blog/2023-12-01-task-lifecycle.markdown

```diff
@@ -0,0 +1,470 @@
+		if (!ret)
+			return 0;
+		panic("Requested init %s failed (error %d).",
+		      execute_command, ret);
+	}
+
+	if (CONFIG_DEFAULT_INIT[0] != '\0') {
+		ret = run_init_process(CONFIG_DEFAULT_INIT);
+		if (ret)
+			pr_err("Default init %s failed (error %d)\n",
+			       CONFIG_DEFAULT_INIT, ret);
+		else
+			return 0;
+	}
+
+	if (!try_to_run_init_process("/sbin/init") ||
+	    !try_to_run_init_process("/etc/init") ||
+	    !try_to_run_init_process("/bin/init") ||
+	    !try_to_run_init_process("/bin/sh"))
+		return 0;
+
+	panic("No working init found.  Try passing init= option to kernel. "
+	      "See Linux Documentation/admin-guide/init.rst for guidance.");
+```
+
+There are several locations it will attempt, in order:
+
+1. Ramdisk file set by `rdinit=` boot command line parameter, with default path
+   `/init`.
+2. File set by `init=` boot command line parameter.
+3. `/sbin/init`
+4. `/etc/init`
+5. `/bin/init`
+6. `/bin/sh`
+
+Should none of these work, the kernel just panics. Which seems fair.
+
+
+## Aside: secondary processors
+
+Until now we've focused on the boot CPU. While the utility of a task still
+applies to a uniprocessor system (perhaps even more so than one with hardware
+parallelism), a nice benefit of encapsulating all the execution state into a
+data structure is being able to load the task onto any other compatible
+processor on the system. But before we can start scheduling on other CPU cores,
+we need to bring them online and initialise them to a state ready for the
+scheduler.
+
+On a pSeries platform, the secondary CPUs are held by the firmware until
+explicitly released by the guest. During boot the boot CPU will, as part of
+`prom_init()`, call `prom_hold_cpus()`. This iterates the list of held secondary
+processors and releases them one by one to the `__secondary_hold` function in
+`arch/powerpc/kernel/head_64.S`. As each starts executing `__secondary_hold`, it
+writes a value to the `__secondary_hold_acknowledge` variable that the boot CPU
+is watching. The secondary processor then immediately starts spinning on
+`__secondary_hold_spinloop`, waiting for it to become non-zero, while the boot
+CPU moves on to the the next processor.
+
+Once every coprocessor is confirmed to be spinning on
+`__secondary_hold_spinloop`, the boot CPU continues on with its boot sequence.
+Eventually it reaches `smp_release_cpus()` early in `start_kernel()` through
+`setup_arch()`, which writes the desired entry point address to
+`__secondary_hold_spinloop`. All the spinning coprocessors now see this value,
+and jump to it. This function, `generic_secondary_smp_init()`, will set up the
+coprocessor's PACA value, perform some machine specific initialisation if
+`cur_cpu_spec->cpu_restore` is set[^4], atomically decrement a
+`spinning_secondaries` variable, and start spinning once again until further
+notice. This time it is waiting on the PACA field `cpu_start`, so we can start
+coprocessors individually.
```

Comment:
May wish to provide a footnote detailing for a wider audience of users that `setup_arch()` stands for 'setup architecture', and that the PACA is Power-specific construct.

Reply:
added a note in the intro